xtractmime/main.py

```diff
@@ -0,0 +1,90 @@
+from typing import List, Union
+
+_APACHE_TYPES_STR = [
+    "text/plain",
+    "text/plain; charset=ISO-8859-1",
+    "text/plain; charset=iso-8859-1",
+    "text/plain; charset=UTF-8",
+]
+
+_APACHE_TYPES_BYTES = [
+    "746578742f706c61696e",
+    "746578742f706c61696e3b20636861727365743d49534f2d383835392d31",
+    "746578742f706c61696e3b20636861727365743d69736f2d383835392d31",
+    "746578742f706c61696e3b20636861727365743d5554462d38",
+]
+
+
+_WHITESPACE_BYTES = [b"\t", b"\r", b"\x0c", b"\n", b" "]
+
+
+# handling resource metadata
+def _should_check_for_apache_bug(supplied_type, http_origin):
+    if http_origin and (
+        supplied_type in _APACHE_TYPES_STR or supplied_type in _APACHE_TYPES_BYTES
+    ):
+        return True
+
+    return False
+
+
+# reading resource header
+def _read_resource_header(body):
+    buffer = b""
+    if len(body) < 1445:
+        buffer = body
+    else:
+        buffer = body[:1445]
+    return buffer
+
+
+# Matching a MIME type pattern
+def _is_match_mime_pattern(
+    input_bytes, byte_pattern, pattern_mask, lead_whitespace=None
+):
+    input_size, pattern_size, mask_size = (
+        len(input_bytes),
+        len(byte_pattern),
+        len(pattern_mask),
+    )
+
+    assert pattern_size == mask_size, "pattern's length should match mask's length"
```

Comment:
Because asserts are removed when Python code is executed in optimized mode, they should only be used to inform developers of assumptions they can make that may not be obvious. For example, here this assert would tell me that, even though it is not obvious from just reading this function, we know that this assertion will always be true because we have control over the possible combinations of `byte_pattern` and `pattern_mask` values received as input.

While this is true for the current implementation of the library, the specification opens the door for additional, user-defined table rows:

> User agents may implicitly extend this table to support additional MIME types.
> — Section 7.1, step 2

So, since we will eventually support user-defined rows, I would instead raise a `ValueError` if these sizes are not the same, which will always be raised, regardless of the Python optimization level.

Also, I’ve just realized that, although [we discussed this point](https://github.com/scrapy/xtractmime/issues/1#issuecomment-855679576), the current API does not allow for this. I believe we will need an additional parameter for the main function to allow user-defined table rows (`extra_types`?). We can leave that for last, but we should not forget about it.

Reply:
I have added the `extra_types` parameter but I will work on it later.

Reply:
Thankyou for the info, I have added `ValueError` and also a test for it.